synapse/lib/crypto/vault.py

```diff
@@ -0,0 +1,516 @@
+            return None
+
+        rkey = self.getCertKey(iden)
+        return Cert.load(byts, rkey=rkey)
+
+    def getCertKey(self, iden):
+        '''
+        Get the RSA Private Key for a given iden.
+
+        Args:
+            iden (str): Iden to retrieve
+
+        Returns:
+            s_rsa.PriKey: The RSA Private Key object.
+        '''
+        byts = self.certkeys.get(iden)
+        if byts is not None:
+            return s_rsa.PriKey.load(byts)
+
+    def genRootCert(self):
+        '''
+        Get or generate the primary root cert for this vault.
+
+        Returns:
+            Cert: A cert helper object
+        '''
+        iden = self.info.get('root')
+        if iden is not None:
+            return self.getCert(iden)
+
+        rkey = self.genRsaKey()
+        tokn = self.genCertTokn(rkey.public())
+        cert = self.genToknCert(tokn, rkey=rkey)
+
+        iden = cert.iden()
+
+        self.info.set('root', iden)
+
+        self.certkeys.set(iden, rkey.save())
+
+        self.addRootCert(cert)
+        return cert
+
+    def addRootCert(self, cert):
+        '''
+        Add a certificate to the Vault as a root certificate.
+
+        Args:
+            cert (Cert): Certificate to add to the Vault.
+
+        Returns:
+            None
+        '''
+        iden = cert.iden()
+        self.roots.set(iden, True)
+        self.certs.set(iden, cert.save())
+
+    def delRootCert(self, cert):
+        '''
+        Delete a root certificate from the Vault.
+
+        Args:
+            cert (Cert): Certificate for the root CA to remove.
+
+        Returns:
+            None
+        '''
+        iden = cert.iden()
+        self.roots.set(iden, False)
+
+    def getRootCerts(self):
+        '''
+        Get a list of the root certificates from the Vault.
+
+        Returns:
+            list: A list of root certificates as Cert objects.
+        '''
+        retn = []
+        for iden, isok in self.roots.items():
+
+            if not isok:
+                continue
+
+            cert = self.getCert(iden)
+            if cert is None:
+                continue
+
+            retn.append(cert)
+
+        return retn
+
+    def isValidCert(self, cert):
+        '''
+        Check if a Vault can validate a Cert against its root certificates.
+
+        Args:
+            cert (Cert): Cert to check against.
+
+        Returns:
+            bool: True if the certificate is valid, False otherwise.
+        '''
+        return any([c.signed(cert) for c in self.getRootCerts()])
+
+@contextlib.contextmanager
+def shared(path):
+    '''
+    A context manager for locking a potentially shared vault.
+
+    Args:
+        path (str): Path to the vault.
+
+    Example:
+
+        with s_vault.shared('~/.syn/vault') as vault:
+            dostuff()
+
+    Yields:
+        Vault: A Vault object.
+    '''
+    full = s_common.genpath(path)
+    lock = os.path.join(full, 'synapse.lock')
+
+    with s_common.lockfile(lock):
```

Comment:
Intraprocess, the ``lockfile()`` does not work.  Do we want to add a module-local ``threading.rlock()`` to prevent multiple threads within a given process from accessing the Vault at once?  Otherwise this could lead to issues with multiple readers having open xacts on the db and cause lmdb to go :boom:  (http://www.openldap.org/lists/openldap-devel/201409/msg00001.html)

Reply:
After discussion: We decided to punt on this at the moment